### PR TITLE
Sync deps and engines with npm

### LINKED ISSUE #2379: Update "tap" and "standard" to newer version
Node-gyp uses to old tap `12.7.0`. In fact now last tap version is `15.x.x`
Node-gyp uses to old standard package `14.3.4`. In fact now last tap version is `16.x.x`

### LINKED ISSUE #2753: [QUESTION] Engines of next major version?
In `npm/cli` we depend on `node-gyp` and would like to make sure the next major version is compatible with `npm` when it comes to `engines.node`.

Previously, `npm` used `^12.13` as the lower base and `node-gyp` used `^12.22` which caused the following issue: https://github.com/npm/cli/issues/4987

`npm@9` was recently released (but is not yet tagged `latest`) with the engines set to `^14.17.0 || ^16.13.0 || >=18.0.0` (see our reasoning here https://github.com/npm/statusboard/issues/519).

Are there plans for a major release soon? If so, has `engines.node` support been discussed?

### LINKED ISSUE #2756: Update `make-fetch-happen`
`make-fetch-happen` dependency needs an update to _v11_ because `@npmcli/move-file` (dependency of  `cacache`) is deprecated.

## Steps to reproduce
- `yarn init -y`
- `yarn set version canary`
- `yarn add node-gyp@9.3.0`

<details><summary>Output</summary>

```console
➤ YN0000: ┌ Resolution step
➤ YN0061: │ @npmcli/move-file@npm:2.0.1 is deprecated: This functionality has been moved to @npmcli/fs
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: Done with warnings
```

</details>

### PR DESCRIPTION
**Pull request by lukekarrys**
##### Checklist

- [x] `npm install && npm test` passes
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)

##### Description of change

The main purpose of this PR is to update dependencies in order to dedupe as many deps as possible within `npm`. As can be seen [in the `npm/cli` source](https://github.com/npm/cli/tree/7018b3d46e10ea4d9d81a478dbdf114b6505ed36/node_modules/node-gyp/node_modules) there are a number of dependencies that end up nested within `node-gyp/node_modules` due to conflicting versions.

To accomplish this a number of other changes were made:

- [`fe5b13a`](https://github.com/nodejs/node-gyp/pull/2770/commits/fe5b13a) **BREAKING CHANGE** update `engines.node` to `^14.17.0 || ^16.13.0 || >=18.0.0` 
- [`a729c15`](https://github.com/nodejs/node-gyp/pull/2770/commits/a729c15) Update `standard` to work with newer JS syntax and fix new lint errors
- [`9ccc66d`](https://github.com/nodejs/node-gyp/pull/2770/commits/9ccc66d) The @npm/cli-team team will be deprecating `npmlog` soon, so this PR drops that dependency in favor of `proc-log` + an otherwise dependency-less custom logger that includes the subset of functionality from `npmlog` used by `node-gyp`
- [`d485f1b`](https://github.com/nodejs/node-gyp/pull/2770/commits/85b642b) Promisifies the `build` command since `which@3.0.0` is promise only
- [`02fc75d`](https://github.com/nodejs/node-gyp/pull/2770/commits/e33ee18) Promisifies the `remove` and `clean` commands after dropping `rimraf` in favor of native recursive rm with `fs.promises.rm`

##### Commits

Each commit in the PR is a separate unit of work and I'd be happy to break it up if this is too difficult to review as is.

##### Reference issues/PRs

After doing the above, I looked through the list of open issues and PRs and identified the following that would either be closed or superseded by this PR.

Closes #2753
Closes #2647
Closes #2358
Closes #2357
Closes #2756
Closes #2379
Closes #2225

### CODE REVIEW

**Comment by lukekarrys** [comment]
#2772 addresses the test failures which are unrelated to the changes in this PR

**Comment by lukekarrys** [comment]
Also if this PR gets accepted, my other PR (#2771) should be reviewed first as I'd want it to make it into the `node-gyp@9` release.

**Comment by lukekarrys** [comment]
@nodejs/node-gyp friendly ping to review this if possible. there's no timeframe, but this refactor will eliminate all non-deduped dependencies within `npm`, so it's a change we're excited to ship

**Review by ruyadorno — APPROVED** [review]

**Comment by lukekarrys** [comment]
Rebased against `main` to resolve conflicts.

**Comment by rvagg** [comment]
ooof, this is a _big_ changeset, kwhich makes it hard to review; I suspect you're going to have to be very patient trying to get someone to review this, I certainly don't have time to look at this in the immediate future unfortunately. Nice to see all of this done, but doing it all in one go means that the chances of getting it in in a decent timeframe are quite low.

**Comment by lukekarrys** [comment]
@rvagg i appreciate the comment and agree that the size of the diff is very big.

i can break it up into multiple PRs and will consider the best way to do that in order to make reviewing easier. if you have any thoughts on what you would like to see there, please let me know.

off the top of my head and i could do the dev deps and linting separately since those changes are 90% automated. and the biggest functional difference is the logging which i could break out as well.

the biggest sticking point is that some of the PRs will need to be stacked since for example the logging change can’t be linked with the current version of `standard`. but if i go that route, i’ll make it clear in what order they should be reviewed.

**Comment by ruyadorno** [comment]
Also if it helps you @rvagg I actually took the time to review the changes and it all looks good to me. While I'm not an owner in this repo, hopefully having another collaborator pair of eyes might help.

**Comment by lukekarrys** [comment]
i'll also add that the commits are ordered and separated very closely to how i would separate them into different PRs, and the ones i linked in the body are the most relevant to a reviewer. but i understand that the github ui leaves something to be desired when reviewing individual commits within a PR

File: lib/clean.js [hunk 1]
```diff
@@ -1,15 +1,17 @@
 'use strict'
 
-const rm = require('rimraf')
-const log = require('npmlog')
+const fs = require('fs')
```
Comment (imatlopez) [inline lib/clean.js:3]: Iirc rimraf is used over fs because it is better cross arch?

File: lib/clean.js [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 
-function clean (gyp, argv, callback) {
+async function clean (gyp, argv) {
   // Remove the 'build' dir
-  var buildDir = 'build'
+  const buildDir = 'build'
 
   log.verbose('clean', 'removing "%s" directory', buildDir)
-  rm(buildDir, callback)
+  await fs.promises.rm(buildDir, { recursive: true, force: true })
```
Comment (imatlopez) [inline lib/clean.js:11]: `require('fs/promises')` would help that

File: lib/remove.js [hunk 3]
```diff
@@ -1,46 +1,45 @@
 'use strict'
 
-const fs = require('fs')
-const rm = require('rimraf')
+const fs = require('fs').promises
```
Comment (imatlopez) [inline lib/remove.js:3]: `require('fs/promises')`

Comment (imatlopez) [reply lib/remove.js:3]: or maybe 'graceful-fs'?

**Comment by imatlopez** [comment]
@rvagg plans for this? Happy it takes over my promisify prs

File: lib/clean.js (hunk 1 above)
Comment (lukekarrys) [reply lib/clean.js:3]: fwiw, `npm` has moved from `rimraf` to `fs/promises` and not seen issues across architectures/platforms. happy to keep `rimraf` in here though if that makes sense

Comment (imatlopez) [reply lib/clean.js:3]: by all means no, less deps the better :)

**Comment by KoenDG** [comment]
Sorry to suddenly interrupt, but this PR seems stuck, and a security update here depends on it: https://github.com/nodejs/node-gyp/pull/2796#issuecomment-1484724721

Reading https://github.com/lerna/lerna/issues/3605 indicates waiting for review, but scrolling up a but shows code comments that make it look like this review has already happened.

Is github not showing everything?

Regardless, what would be required for this to progress? I'm asking because it's related to a security issue, which if I'm reading it right cannot be resolved until this PR is finished and merged: https://nvd.nist.gov/vuln/detail/CVE-2022-25881

Apologies if this comes off strong, it's not my intent. It's just something I noticed a while back and it seems there's no movement. Thought I'd shake the grapevine a bit. Is there someone who needs to get involved in this that has perhaps lost track of it? It happens.

**Comment by rvagg** [comment]
Hey @imatlopez are you able to help review this? It really needs someone with the time and experience with this codebase to walk through it and give it the :thumbsup:. I'm unlikely to be able to find any time to do this but on the surface it seems like a worthwhile set of changes, just too much.

**Comment by rvagg** [comment]
I should acknowledge that @ruyadorno also walked through this, that's helpful. It would be good if we had more than one person who's name's familiar around here to give this a tick. This code goes out to a _lot_ of people so we can't ship big changes lightly.

File: lib/log.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    yellow: 43,
+    blue: 44,
+    magenta: 45,
+    cyan: 46,
+    white: 47
+  }
+}
+
+class Logger {
+  #buffer = []
```
Comment (imatlopez) [inline lib/log.js:43]: observation: min engine is v12 in package.json, which is when private fields were added to node.

File: lib/process-release.js [hunk 5]
```diff
@@ -1,11 +1,11 @@
-/* eslint-disable node/no-deprecated-api */
+/* eslint-disable n/no-deprecated-api */
```
Comment (imatlopez) [inline lib/process-release.js:1]: guess this line is not needed if this didn't cause issues?

**Comment by imatlopez** [comment]
@lukekarrys you have conflicts in the pr, @rvagg gave it another pass today and besides those two comment, it looked ok to me.

**Comment by lukekarrys** [comment]
I will resolve the conflicts and address the comments this week.

File: lib/process-release.js (hunk 5 above)
Comment (lukekarrys) [reply lib/process-release.js:1]: The update of `standard` to the latest version uses a new eslint plugin called `eslint-plugin-n` (https://www.npmjs.com/package/eslint-plugin-n). So disabling this rule had to change to the plugin name.

File: lib/remove.js (hunk 3 above)
Comment (lukekarrys) [reply lib/remove.js]: This has been addresses by changing to `fs/promises` in all relevant files.

File: lib/clean.js (hunk 2 above)
Comment (lukekarrys) [reply lib/clean.js]: This has been updated to `fs/promises`.

**Comment by lukekarrys** [comment]
This PR has been rebased to fix conflicts with the latest commits in `main`. Since `make-fetch-happen` was updated in #2796, I dropped the commit updating that from this PR.

**Comment by lukekarrys** [comment]
@imatlopez @rvagg Thank you both for your feedback on this PR. I believe I've addressed all the feedback presented so far and fix the conflicts.

CI was previously failing on what I believe was a flaky test, but I will wait and watch the currently running CI on this PR to ensure it all passes.

**Review by imatlopez — APPROVED** [review]

**Comment by JamesHenry** [comment]
Thanks for all your work on this @lukekarrys!

Looking forward to this landing!

**Comment by imatlopez** [comment]
@lukekarrys seems this has new conflicts @rvagg once those are resolved is merge on the table?

**Comment by lukekarrys** [comment]
Thanks for the heads up @imatlopez! I rebased this PR to fix all the conflicts. Note that the original body of the PR is also updated with the latest commit links.

The only real change to fix the conflicts was to drop the commit updating `tap` since this uses `mocha` now as a test runner.

**Comment by imatlopez** [comment]
I'd imagine `tap` would be preferred now that it comes with node but long live mocha?

2 windows tests are failing (but not all so flaky?)

**Comment by cclauss** [comment]
Re-running the two failed visual studio jobs.

**Comment by JamesHenry** [comment]
@cclauss Thanks a lot for rerunning, looks like this is good to go now?

(This issue indirectly causes a warning on every fresh lerna repo, that's why I'm particularly invested in its outcome 😄)

**Comment by cclauss** [comment]
@JamesHenry Please `Review changes` at https://github.com/nodejs/node-gyp/pull/2770/files

**Comment by lukekarrys** [comment]
The plan is to release this as v10.0.0 sometime in the next couple weeks. Thanks for your patience @JamesHenry.

**Pull request merged**